qcodes/instrument_drivers/QuantumDesign/DynaCoolPPMS/DynaCool.py

```diff
@@ -0,0 +1,281 @@
+    def error_code(self):
+        return self._error_code
+
+    @staticmethod
+    def _pick_one(which_one: int, parser: type, resp: str) -> Any:
+        """
+        Since most of the API calls return several values in a comma-seperated
+        string, here's a convenience function to pick out the substring of
+        interest
+        """
+        return parser(resp.split(', ')[which_one])
+
+    def get_idn(self) -> Dict[str, Optional[str]]:
+        response = self.ask('*IDN?')
+        # just clip out the error code
+        idparts = response[2:].split(', ')
+
+        return dict(zip(('vendor', 'model', 'serial', 'firmware'), idparts))
+
+    def _present_field_getter(self) -> float:
+        resp = self.ask('FELD?')
+        number_in_oersted = cast(float, DynaCool._pick_one(1, float, resp))
+        number_in_SI = number_in_oersted/79.57747
+        return number_in_SI
+
+    def _field_getter(self, param_name: str) -> Union[int, float]:
+        """
+        The combined get function for the three field parameters,
+        field_setpoint, field_rate, and field_approach
+        """
+        raw_response = self.ask('GLFS?')
+        sp = DynaCool._pick_one(1, float, raw_response)
+        rate = DynaCool._pick_one(2, float, raw_response)
+        approach = DynaCool._pick_one(3, int, raw_response)
+
+        return dict(zip(self.field_params, [sp, rate, approach]))[param_name]
+
+    def _field_setter(self, param: str, value: float) -> None:
+        """
+        The combined set function for the three field parameters,
+        field_setpoint, field_rate, and field_approach
+        """
+        vals: List[Union[int, float]]
+        vals = list(self.parameters[p].raw_value for p in self.field_params)
+        vals[self.field_params.index(param)] = value
+
+        self.write(f'FELD {vals[0]}, {vals[1]}, {vals[2]}, 0')
+
+    def _temp_getter(self, param_name: str) -> Union[int, float]:
+        """
+        This function queries the last temperature setpoint (w. rate and mode)
+        from the instrument.
+        """
+        raw_response = self.ask('GLTS?')
+        sp = DynaCool._pick_one(1, float, raw_response)
+        rate = DynaCool._pick_one(2, float, raw_response)
+        mode = DynaCool._pick_one(3, int, raw_response)
+
+        return dict(zip(self.temp_params, [sp, rate, mode]))[param_name]
+
+    def _temp_setter(self, param: str, value: float) -> None:
+        """
+        The setter function for the temperature parameters. All three are set
+        with the same call to the instrument API
+        """
+        vals: List[Union[int, float]]
+        vals = list(self.parameters[par].raw_value for par in self.temp_params)
+        vals[self.temp_params.index(param)] = value
+
+        self.write(f'TEMP {vals[0]}, {vals[1]}, {vals[2]}')
+
+    def write(self, cmd: str) -> None:
+        """
+        Since the error code is always returned, we must read it back
+        """
+        super().write(cmd)
```

Comment:
Maybe the line below should be logged to the visa_log @Dominik-Vogel knows more

Reply:
Exactly! The `super.write(cmd)` does already the logging to the visa log, but the following reading of the status code should also log to the visa log via `self.visa_log`

Reply:
Done.